internal/db/transactions.go

```diff
@@ -0,0 +1,45 @@
+package db
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+)
+
+// withTransaction creates a new transaction and handles rollback/commit based on the
+// error object returned by the `TxFn` or when it panics.
+func (db *dB) withTransaction(ctx context.Context, fn func(tx pgx.Tx) error) error {
+	tx, beginErr := db.Pool.Begin(ctx)
+	if beginErr != nil {
+		return fmt.Errorf("tx error: %w", beginErr)
+	}
+
+	defer func() {
+		if p := recover(); p != nil {
+			rollErr := tx.Rollback(ctx)
+			if rollErr != nil {
+				return
+			}
+			panic(p)
```

Comment:
This is a standard Go idiom when you use `recover`. Basically, we want to do an extra action (rollback) and then keep panicking. This code is only used when panic occurs within the block (e.g. nil pointer error), it will *not* be used for regular rollbacks caused by returning an error from the block.